docs/proposals/0001-cli-design.md

```diff
@@ -0,0 +1,49 @@
+# CLI design
+
+authors: @toumorokoshi, @zph
+
+
+This proposal outlines the design decisions and considerations for the Tome CLI.
+
+## Background
+
+Tome was initially designed as a drop-in replacement for [sub](https://github.com/basecamp/sub), to ensure portability.
+
+## Terminology
+
+For the following discussion, `{instance}` is defined as the instantiated tome or sub command.
+
+## Design Decisions
+
+### compatibility with sub CLI instances
+
+Sub-based clis provide the following functionality with each command:
+
+- `{instance} --help`: a list of all known commands, and a description.
+- `{instance} help {subcommand}`: a help guide on the subcommand.
+
+### reserving a namespace for tome to add to instanced CLIs
+
+Since a user can name their commands after any valid filename, it is possible
+for any command that is built into the tome-based CLIs to clobber a possible command the user would like to create.
+
+For example let's assume that tome would like to introduce a top-level command known as "exec". This would look like:
```

Comment:
Looking at prior art, sub solved this by setting up a few specific builtins:

https://github.com/basecamp/sub#whats-on-your-sub

```
You get a few commands that come with your sub:

commands: Prints out every subcommand available
completions: Helps kick off subcommand autocompletion.
help: Document how to use each subcommand
init: Shows how to load your sub with autocompletions, based on your shell.
shell: Helps with calling subcommands that might be named the same as builtin/executables.
```

Reply:
Tome's architecture is a bit different because some of those commands are commands to the bare binary where we have full liberty to determine subcommands without naming conflict:

* init
* completions
* shell

While others are commands to the `{instance}`:

* help
* commands

So we could declare `help and commands` as reserved names for `{instance}` and then design a bare tome to handle init, completions, shell. Furthermore if we're concerned of later conflicting names between our `{instance}` arguments that are builtin vs user supplied, we could pre-declare that anything in the `{instance} tome ....` namespace is reserved. I don't see a looming reason to pre-stake it but doesn't do us any harm to declare that interface and reservation now :).

In #5 I setup the following subcommands to the bare tome:

```
SUBCOMMANDS:
    commands    List available scripts
    complete    Output commandline autocompletion results
    exec        Excute script
    help        Prints this message or the help of the given subcommand(s)
    init        Print shell completion
    init_v2     Print shell completion
```

My preference is we keep `exec` instead of `shell` because of naming clarity. But we can rename `complete` to `completions` to match how sub operates.

Reply:
+1 to everything you stated here. Specifically:

- huge +1 to pre-reserving the `tome` namespace, just in case.
- no strong opinion regarding how the commands moved to the binary should be named (e.g. complete vs completions). It shouldn't impact those who are trying to use it. 

Would you be interested in directly adding these as suggestions? I'm not clear on how easy that is via GitHub... otherwise I'll try to summarize your points and add them.

Reply:
@zph replying to your comment in this thread:

> thinking Alternatively, since we'll rarely extend the interface of {instance} we add a third instance reserved name called exec which could be used to avoid duplicate naming:

Yes, that makes sense, and we should probably have that one last escape hatch.

Or it could be a good use for the tome namespace:

```
{instance} tome exec {your-command-here}
```

re-reading [what's on your sub](https://github.com/basecamp/sub#whats-on-your-sub), the `shell` subcommand is an interesting one: I think it also existed to try to defend against a filename that matches a posix builtin (e.g. `env`)?

I think that won't happen with tome's design, since we always call the full path of the executable, and it immediately disambiguates the command executed. I suspect sub had to add this since it cds around into peoples directories and ends up calling the file name directly.

Reply:
@toumorokoshi  I'm not sure on the permissioning, but I'll try to pull down the branch, make these suggestions and push it back up for review.